### PR TITLE
Generic helper for consistent post-reconcile spec and status updates

### PR DESCRIPTION
**Pull request by csams**
## Summary
Unify post-reconcile spec and status updating logic in a type safe way. The `APIBinding` and `APIExport` controllers are updated as examples.

### CODE REVIEW

File: pkg/reconciler/apis/apibinding/apibinding_controller.go [hunk 1]
```diff
@@ -129,6 +126,7 @@ func NewController(
 		},
 		crdIndexer:        crdInformer.Informer().GetIndexer(),
 		deletedCRDTracker: newLockedStringSet(),
+		postReconcile:     postreconcile.NewPostReconcile[*APIBinding, *APIBindingSpec, *APIBindingStatus](kcpClusterClient.ApisV1alpha1().APIBindings()),
```
Comment (sttts) [inline pkg/reconciler/apis/apibinding/apibinding_controller.go:129]: isn't that more a "commit" phase?

**Comment by sttts** [comment]
Sgtm.

Only bike-shedding: postreconcile => commit.

File: pkg/reconciler/apis/apibinding/apibinding_controller.go (hunk 1 above)
Comment (csams) [reply pkg/reconciler/apis/apibinding/apibinding_controller.go]: Makes sense. I updated the naming.

File: pkg/reconciler/committer/committer.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+		}
+
+		logger.V(2).Info(fmt.Sprintf("patching %s", focusType), "patch", string(patchBytes))
+		_, err = patcher.Patch(logicalcluster.WithCluster(ctx, clusterName), obj.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, subresources...)
+		if err != nil {
+			return fmt.Errorf("failed to patch %s %s|%s: %w", focusType, clusterName, name, err)
+		}
+
+		// Simultaneous updates of spec and status are never allowed.
+		if specOrObjectMetaChanged && statusChanged {
```
Comment (csams) [inline pkg/reconciler/committer/committer.go:115]: @stevekuznetsov Shouldn't this go near the top?

Comment (stevekuznetsov) [reply pkg/reconciler/committer/committer.go:115]: yes, that should be as early as possible

**Pull request merged**